gtsfm/utils/io.py

```diff
@@ -162,4 +175,44 @@ def write_images(gtsfm_data: GtsfmData, save_dir: str) -> None:
             tx, ty, tz = wti
             qw, qx, qy, qz = wRi_quaternion
 
-            f.write(f"{i} {qw} {qx} {qy} {qz} {tx} {ty} {tz}\n")
+            f.write(f"{i} {qw} {qx} {qy} {qz} {tx} {ty} {tz} {i} {img_fname}\n")
+            # TODO: write out the points2d
+
+
+def write_points(gtsfm_data: GtsfmData, save_dir: str) -> None:
+    """Writes the point cloud data file in the COLMAP format.
+
+    Args:
+        gtsfm_data: scene data to write.
+        save_dir: folder to put the points3D.txt file in.
+    """
+    os.makedirs(save_dir, exist_ok=True)
+
+    num_pts = gtsfm_data.number_tracks()
+    avg_track_length, _, _ = gtsfm_data.get_track_length_statistics()
+
+    file_path = os.path.join(save_dir, "points3D.txt")
+    with open(file_path, "w") as f:
+        f.write("# Number of cameras: {}\n".format(gtsfm_data.number_images()))
+
+        f.write("# 3D point list with one line of data per point:\n")
+        f.write("#   POINT3D_ID, X, Y, Z, R, G, B, ERROR, TRACK[] as (IMAGE_ID, POINT2D_IDX)\n")
+        f.write(f"# Number of points: {num_pts}, mean track length: {avg_track_length}\n")
+
+        # TODO: extract point color
+        r, g, b = 0, 0, 0
+        # TODO: coming in another PR
+        track_reproj_error = 0
+        # TODO: assign unique indices to all keypoints (2d points)
+        point2d_idx = 0
```

Comment:
These indices will be 0-n per image, right?

Reply:
I don't think so.

POINT2D_IDX defines the zero-based index of the keypoint in the images.txt file

https://colmap.github.io/format.html#points3d-txt

Reply:
Actually, maybe you are right. It's a bit unclear.

Reply:
@adisingh50 would you mind looking into this issue here about what the 2d keypoint/point index means?